File: mobile/ios/Sources/YuleDKCS/BLEManager.swift
import Foundation
import CoreBluetooth
import Combine

/// BLE Manager for vehicle communication
///
/// Handles Bluetooth Low Energy connections, command sending, scanning,
/// and real-time status monitoring for vehicles.
public class BLEManager: NSObject, ObservableObject {
    
    public static let shared = BLEManager()
    
    // MARK: - Published Properties
    
    @Published public private(set) var scanResults: [ScanResult] = []
    @Published public private(set) var isScanning = false
    @Published public private(set) var connectionState: ConnectionState = .disconnected
    @Published public private(set) var vehicleStatus: VehicleStatus?
    
    // MARK: - BLE UUIDs
    
    public static let serviceUUID = CBUUID(string: "0000yule-0000-1000-8000-00805f9b34fb")
    public static let commandCharUUID = CBUUID(string: "0000cmd1-0000-1000-8000-00805f9b34fb")
    public static let responseCharUUID = CBUUID(string: "0000resp-0000-1000-8000-00805f9b34fb")
    public static let statusCharUUID = CBUUID(string: "0000stat-0000-1000-8000-00805f9b34fb")
    
    // MARK: - Command Codes
    
    public enum CommandCode: UInt8 {
        case lock = 0x01
        case unlock = 0x02
        case startEngine = 0x03
        case stopEngine = 0x04
        case openTrunk = 0x05
        case openWindows = 0x06
        case closeWindows = 0x07
        case findVehicle = 0x08
        case getStatus = 0x09
    }
    
    // MARK: - Connection State
    
    public enum ConnectionState: Equatable {
        case disconnected
        case scanning
        case connecting
        case connected(String)
        case error(String)
    }
    
    // MARK: - Private Properties
    
    private var centralManager: CBCentralManager!
    private var connectedPeripheral: CBPeripheral?
    private var commandCharacteristic: CBCharacteristic?
    private var responseCharacteristic: CBCharacteristic?
    private var statusCharacteristic: CBCharacteristic?
    
    private var pendingCommands: [String: (Result<Data, Error>) -> Void] = [:]
    private var scanTimer: Timer?
    
    private var cancellables = Set<AnyCancellable>()
    private let queue = DispatchQueue(label: "com.yuledkcs.ble", qos: .userInitiated)
    
    // MARK: - Initialization
    
    private override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: queue)
    }
    
    // MARK: - Scanning
    
    /// Start scanning for nearby vehicles
    /// - Parameters:
    ///   - timeout: Scan timeout in seconds
    public func startScan(timeout: TimeInterval = 10.0) {
        guard centralManager.state == .poweredOn else {
            connectionState = .error("蓝牙未开启")
            return
        }
        
        isScanning = true
        connectionState = .scanning
        scanResults = []
        
        centralManager.scanForPeripherals(
            withServices: [BLEManager.serviceUUID],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )
        
        // Auto stop after timeout
        scanTimer?.invalidate()
        scanTimer = Timer.scheduledTimer(withTimeInterval: timeout, repeats: false) { [weak self] _ in
            self?.stopScan()
        }
    }
    
    /// Stop scanning for devices
    public func stopScan() {
        centralManager.stopScan()
        isScanning = false
        scanTimer?.invalidate()
        scanTimer = nil
        
        if case .scanning = connectionState {
            connectionState = .disconnected
        }
    }
    
    // MARK: - Connection
    
    /// Connect to a vehicle via BLE
    /// - Parameters:
    ///   - address: UUID of the BLE device
    ///   - autoReconnect: Enable auto-reconnect
    public func connect(to address: UUID, autoReconnect: Bool = true) {
        stopScan()
        
        guard let peripheral = centralManager.retrievePeripherals(withIdentifiers: [address]).first else {
            connectionState = .error("设备未找到")
            return
        }
        
        connectionState = .connecting
        connectedPeripheral = peripheral
        peripheral.delegate = self
        
        centralManager.connect(peripheral, options: [
            CBConnectPeripheralOptionNotifyOnConnectionKey: true,
            CBConnectPeripheralOptionNotifyOnDisconnectionKey: true
        ])
    }
    
    /// Connect to nearest vehicle
    public func connectToNearest(completion: @escaping (Result<String, Error>) -> Void) {
        guard let nearest = scanResults.first else {
            completion(.failure(YuleDKCSError.bleNotAvailable))
            return
        }
        
        connect(to: nearest.peripheral.identifier)
        
        // Wait for connection
        $connectionState
            .filter { state in
                if case .connected = state { return true }
                if case .error = state { return true }
                return false
            }
            .first()
            .sink { state in
                switch state {
                case .connected(let address):
                    completion(.success(address))
                case .error(let message):
                    completion(.failure(YuleDKCSError.connectionFailed))
                default:
                    break
                }
            }
            .store(in: &cancellables)
    }
    
    /// Disconnect from the current device
    public func disconnect() {
        guard let peripheral = connectedPeripheral else { return }
        centralManager.cancelPeripheralConnection(peripheral)
    }
    
    // MARK: - Commands
    
    /// Send a command to the connected vehicle
    public func sendCommand(
        keyId: String,
        command: Command,
        completion: @escaping (Result<CommandResponse, Error>) -> Void
    ) {
        guard let peripheral = connectedPeripheral,
              peripheral.state == .connected else {
            completion(.failure(YuleDKCSError.connectionFailed))
            return
        }
        
        guard let characteristic = commandCharacteristic else {
            completion(.failure(YuleDKCSError.bleNotAvailable))
            return
        }
        
        // Check permission
        guard KeyManager.shared.hasPermission(keyId: keyId, permission: command.toPermission()) else {
            completion(.failure(YuleDKCSError.permissionDenied))
            return
        }
        
        queue.async { [weak self] in
            do {
                let packet = try self?.buildCommandPacket(keyId: keyId, command: command)
                guard let data = packet else { return }
                
                let commandId = UUID().uuidString
                
                // Store completion handler
                self?.pendingCommands[commandId] = { result in
                    DispatchQueue.main.async {
                        switch result {
                        case .success(let responseData):
                            let response = self?.parseResponse(data: responseData) ?? CommandResponse(success: false, message: "解析失败")
                            
                            // Record usage
                            KeyManager.shared.recordUsage(
                                keyId: keyId,
                                operation: command.description,
                                status: response.success ? .success : .failure,
                                failureReason: response.success ? nil : response.message
                            )
                            
                            completion(.success(response))
                            
                        case .failure(let error):
                            KeyManager.shared.recordUsage(
                                keyId: keyId,
                                operation: command.description,
                                status: .failure,
                                failureReason: error.localizedDescription
                            )
                            completion(.failure(error))
                        }
                    }
                }
                
                peripheral.writeValue(data, for: characteristic, type: .withResponse)
                
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }
    
    // MARK: - Convenience Methods
    
    public func lock(keyId: String, completion: @escaping (Result<CommandResponse, Error>) -> Void) {
        sendCommand(keyId: keyId, command: .lock, completion: completion)
    }
    
    public func unlock(keyId: String, completion: @escaping (Result<CommandResponse, Error>) -> Void) {
        sendCommand(keyId: keyId, command: .unlock, completion: completion)
    }
    
    public func startEngine(keyId: String, completion: @escaping (Result<CommandResponse, Error>) -> Void) {
        sendCommand(keyId: keyId, command: .startEngine, completion: completion)
    }
    
    public func stopEngine(keyId: String, completion: @escaping (Result<CommandResponse, Error>) -> Void) {
        sendCommand(keyId: keyId, command: .stopEngine, completion: completion)
    }
    
    public func openTrunk(keyId: String, completion: @escaping (Result<CommandResponse, Error>) -> Void) {
        sendCommand(keyId: keyId, command: .openTrunk, completion: completion)
    }
    
    public func openWindows(keyId: String, completion: @escaping (Result<CommandResponse, Error>) -> Void) {
        sendCommand(keyId: keyId, command: .openWindows, completion: completion)
    }
    
    public func closeWindows(keyId: String, completion: @escaping (Result<CommandResponse, Error>) -> Void) {
        sendCommand(keyId: keyId, command: .closeWindows, completion: completion)
    }
    
    public func findVehicle(keyId: String, completion: @escaping (Result<CommandResponse, Error>) -> Void) {
        sendCommand(keyId: keyId, command: .findVehicle, completion: completion)
    }
    
    /// Request current vehicle status
    public func requestVehicleStatus() {
        guard let peripheral = connectedPeripheral,
              let characteristic = commandCharacteristic else { return }
        
        let data = Data([CommandCode.getStatus.rawValue])
        peripheral.writeValue(data, for: characteristic, type: .withResponse)
    }
    
    // MARK: - Private Methods
    
    private func buildCommandPacket(keyId: String, command: Command) throws -> Data {
        guard let key = KeyManager.shared.getKey(keyId: keyId) else {
            throw YuleDKCSError.keyNotFound
        }
        
        let timestamp = Int64(Date().timeIntervalSince1970)
        
        let commandByte: UInt8 = {
            switch command {
            case .lock: return CommandCode.lock.rawValue
            case .unlock: return CommandCode.unlock.rawValue
            case .startEngine: return CommandCode.startEngine.rawValue
            case .stopEngine: return CommandCode.stopEngine.rawValue
            case .openTrunk: return CommandCode.openTrunk.rawValue
            case .openWindows: return CommandCode.openWindows.rawValue
            case .closeWindows: return CommandCode.closeWindows.rawValue
            case .findVehicle: return CommandCode.findVehicle.rawValue
            case .custom(let data): return data.first ?? 0
            }
        }()
        
        // Build packet: [VERSION][KEY_ID][TIMESTAMP][COMMAND][SIGNATURE]
        var packet = Data()
        packet.append(0x01) // Version
        
        // Key ID (16 bytes)
        let keyIdData = keyId.data(using: .utf8) ?? Data()
        packet.append(keyIdData.prefix(16))
        packet.append(contentsOf: Array(repeating: 0, count: max(0, 16 - keyIdData.count)))
        
        // Timestamp (8 bytes, big-endian)
        var timestampBytes = timestamp.bigEndian
        packet.append(UnsafeBufferPointer(start: &timestampBytes, count: MemoryLayout<Int64>.size))
        
        // Command
        packet.append(commandByte)
        
        // Sign packet
        let dataToSign = packet
        let signature = try NativeLib.signCommand(data: dataToSign, keyData: key.keyData)
        packet.append(signature)
        
        return packet
    }
    
    private func parseResponse(data: Data) -> CommandResponse {
        guard data.count >= 1 else {
            return CommandResponse(success: false, message: "无效响应")
        }
        
        let status = data[0]
        return CommandResponse(
            success: status == 0x00,
            message: status == 0x00 ? "成功" : "失败: \(status)"
        )
    }
    
    private func handleStatusUpdate(data: Data) {
        guard data.count >= 5 else { return }
        
        let statusByte = data[0]
        let battery = data[1]
        let fuel = data[2]
        
        vehicleStatus = VehicleStatus(
            doorLocked: (statusByte & 0x01) != 0,
            engineRunning: (statusByte & 0x02) != 0,
            trunkOpen: (statusByte & 0x04) != 0,
            windowsOpen: (statusByte & 0x08) != 0,
            batteryLevel: Int(battery),
            fuelLevel: Int(fuel),
            alarmActive: (statusByte & 0x10) != 0,
            timestamp: Date().timeIntervalSince1970
        )
    }
}

// MARK: - CBCentralManagerDelegate

extension BLEManager: CBCentralManagerDelegate {
    
    public func centralManagerDidUpdateState(_ central: CBCentralManager) {
        if central.state != .poweredOn {
            connectionState = .error("蓝牙未开启")
        }
    }
    
    public func centralManager(
        _ central: CBCentralManager,
        didDiscover peripheral: CBPeripheral,
        advertisementData: [String: Any],
        rssi RSSI: NSNumber
    ) {
        let result = ScanResult(peripheral: peripheral, rssi: RSSI.intValue)
        
        DispatchQueue.main.async { [weak self] in
            // Update or add result
            if let index = self?.scanResults.firstIndex(where: { $0.peripheral.identifier == peripheral.identifier }) {
                self?.scanResults[index] = result
            } else {
                self?.scanResults.append(result)
            }
            
            // Sort by RSSI
            self?.scanResults.sort { $0.rssi > $1.rssi }
        }
    }
    
    public func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        DispatchQueue.main.async { [weak self] in
            self?.connectionState = .connected(peripheral.identifier.uuidString)
        }
        peripheral.discoverServices([BLEManager.serviceUUID])
    }
    
    public func centralManager(
        _ central: CBCentralManager,
        didFailToConnect peripheral: CBPeripheral,
        error: Error?
    ) {
        DispatchQueue.main.async { [weak self] in
            self?.connectionState = .error(error?.localizedDescription ?? "连接失败")
        }
    }
    
    public func centralManager(
        _ central: CBCentralManager,
        didDisconnectPeripheral peripheral: CBPeripheral,
        error: Error?
    ) {
        DispatchQueue.main.async { [weak self] in
            self?.connectionState = .disconnected
            self?.connectedPeripheral = nil
            self?.vehicleStatus = nil
        }
    }
}

// MARK: - CBPeripheralDelegate

extension BLEManager: CBPeripheralDelegate {
    
    public func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard let services = peripheral.services else { return }
        
        for service in services where service.uuid == BLEManager.serviceUUID {
            peripheral.discoverCharacteristics(
                [BLEManager.commandCharUUID, BLEManager.responseCharUUID, BLEManager.statusCharUUID],
                for: service
            )
        }
    }
    
    public func peripheral(
        _ peripheral: CBPeripheral,
        didDiscoverCharacteristicsFor service: CBService,
        error: Error?
    ) {
        guard let characteristics = service.characteristics else { return }
        
        for characteristic in characteristics {
            switch characteristic.uuid {
            case BLEManager.commandCharUUID:
                commandCharacteristic = characteristic
            case BLEManager.responseCharUUID:
                responseCharacteristic = characteristic
                peripheral.setNotifyValue(true, for: characteristic)
            case BLEManager.statusCharUUID:
                statusCharacteristic = characteristic
                peripheral.setNotifyValue(true, for: characteristic)
            default:
                break
            }
        }
        
        // Request initial status
        requestVehicleStatus()
    }
    
    public func peripheral(
        _ peripheral: CBPeripheral,
        didUpdateValueFor characteristic: CBCharacteristic,
        error: Error?
    ) {
        guard let data = characteristic.value else { return }
        
        switch characteristic.uuid {
        case BLEManager.responseCharUUID:
            handleResponse(data: data)
        case BLEManager.statusCharUUID:
            handleStatusUpdate(data: data)
        default:
            break
        }
    }
    
    private func handleResponse(data: Data) {
        guard data.count >= 2 else { return }
        
        let commandId = data[1]
        let callback = pendingCommands.removeValue(forKey: String(commandId))
        
        if data[0] == 0x00 {
            callback?(.success(data))
        } else {
            callback?(.failure(YuleDKCSError.commandFailed))
        }
    }
}

// MARK: - Supporting Types

public struct ScanResult: Identifiable {
    public let id = UUID()
    public let peripheral: CBPeripheral
    public let rssi: Int
    
    public var name: String? { peripheral.name }
}

public struct CommandResponse {
    public let success: Bool
    public let message: String
}

private extension Command {
    func toPermission() -> Permission {
        switch self {
        case .lock: return .lock
        case .unlock: return .unlock
        case .startEngine: return .startEngine
        case .stopEngine: return .stopEngine
        case .openTrunk: return .openTrunk
        case .openWindows: return .openWindows
        case .closeWindows: return .closeWindows
        case .findVehicle: return .findVehicle
        case .custom: return .unlock
        }
    }
}

File: mobile/android/src/main/java/com/yuledkcs/BLEManager.kt
package com.yuledkcs

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.ParcelUuid
import android.util.Log
import com.yuledkcs.model.Command
import com.yuledkcs.model.VehicleStatus
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import no.nordicsemi.android.ble.BleManager
import no.nordicsemi.android.ble.data.Data
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * BLE Manager for vehicle communication
 * 
 * Handles Bluetooth Low Energy connections, command sending, scanning,
 * and real-time status monitoring for vehicles.
 */
class BLEManager private constructor() : BleManager(YuleDKCS.context) {
    
    companion object {
        private const val TAG = "BLEManager"
        
        // BLE Service and Characteristic UUIDs
        val SERVICE_UUID = UUID.fromString("0000yule-0000-1000-8000-00805f9b34fb")
        val COMMAND_CHAR_UUID = UUID.fromString("0000cmd1-0000-1000-8000-00805f9b34fb")
        val RESPONSE_CHAR_UUID = UUID.fromString("0000resp-0000-1000-8000-00805f9b34fb")
        val STATUS_CHAR_UUID = UUID.fromString("0000stat-0000-1000-8000-00805f9b34fb")
        
        @Volatile
        private var instance: BLEManager? = null
        
        fun getInstance(): BLEManager {
            return instance ?: synchronized(this) {
                instance ?: BLEManager().also { instance = it }
            }
        }
        
        // Command codes
        const val CMD_LOCK = 0x01
        const val CMD_UNLOCK = 0x02
        const val CMD_START_ENGINE = 0x03
        const val CMD_STOP_ENGINE = 0x04
        const val CMD_OPEN_TRUNK = 0x05
        const val CMD_OPEN_WINDOWS = 0x06
        const val CMD_CLOSE_WINDOWS = 0x07
        const val CMD_FIND_VEHICLE = 0x08
        const val CMD_GET_STATUS = 0x09
    }
    
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var commandCharacteristic: BluetoothGattCharacteristic? = null
    private var responseCharacteristic: BluetoothGattCharacteristic? = null
    private var statusCharacteristic: BluetoothGattCharacteristic? = null
    
    private val pendingCommands = ConcurrentHashMap<String, (Result<ByteArray>) -> Unit>()
    private var connectionCallback: ((Boolean) -> Unit)? = null
    private var currentDeviceAddress: String? = null
    
    // Scanning state
    private val _scanResults = MutableStateFlow<List<ScanResult>>(emptyList())
    val scanResults: StateFlow<List<ScanResult>> = _scanResults
    
    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning
    
    // Connection state
    private val _connectionState = MutableStateFlow<ConnectionState>(ConnectionState.Disconnected)
    val connectionState: StateFlow<ConnectionState> = _connectionState
    
    // Vehicle status
    private val _vehicleStatus = MutableStateFlow<VehicleStatus?>(null)
    val vehicleStatus: StateFlow<VehicleStatus?> = _vehicleStatus
    
    private var scanCallback: ScanCallback? = null
    
    sealed class ConnectionState {
        object Disconnected : ConnectionState()
        object Scanning : ConnectionState()
        object Connecting : ConnectionState()
        data class Connected(val address: String) : ConnectionState()
        data class Error(val message: String) : ConnectionState()
    }
    
    /**
     * Start scanning for nearby vehicles
     * 
     * @param timeoutMs Scan timeout in milliseconds (default 10000)
     * @param serviceUuid Optional service UUID filter
     */
    fun startScan(
        timeoutMs: Long = 10000,
        serviceUuid: UUID? = SERVICE_UUID
    ) {
        YuleDKCS.checkInitialized()
        
        val bluetoothAdapter = android.bluetooth.BluetoothAdapter.getDefaultAdapter()
        val scanner = bluetoothAdapter?.bluetoothLeScanner
        
        if (scanner == null) {
            Log.e(TAG, "Bluetooth LE scanner not available")
            return
        }
        
        _isScanning.value = true
        _connectionState.value = ConnectionState.Scanning
        
        // Build scan filter
        val filters = serviceUuid?.let {
            listOf(
                ScanFilter.Builder()
                    .setServiceUuid(ParcelUuid(it))
                    .build()
            )
        } ?: emptyList()
        
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()
        
        scanCallback = object : ScanCallback() {
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                val current = _scanResults.value.toMutableList()
                // Update or add result
                val index = current.indexOfFirst { it.device.address == result.device.address }
                if (index >= 0) {
                    current[index] = result
                } else {
                    current.add(result)
                }
                // Sort by RSSI (signal strength)
                current.sortByDescending { it.rssi }
                _scanResults.value = current
            }
            
            override fun onBatchScanResults(results: List<ScanResult>) {
                _scanResults.value = results.sortedByDescending { it.rssi }
            }
            
            override fun onScanFailed(errorCode: Int) {
                Log.e(TAG, "Scan failed with error: $errorCode")
                _isScanning.value = false
                _connectionState.value = ConnectionState.Error("扫描失败: $errorCode")
            }
        }
        
        scanner.startScan(filters, settings, scanCallback)
        
        // Auto stop after timeout
        scope.launch {
            delay(timeoutMs)
            stopScan()
        }
    }
    
    /**
     * Stop scanning for devices
     */
    fun stopScan() {
        val bluetoothAdapter = android.bluetooth.BluetoothAdapter.getDefaultAdapter()
        val scanner = bluetoothAdapter?.bluetoothLeScanner
        
        scanCallback?.let { callback ->
            scanner?.stopScan(callback)
        }
        
        _isScanning.value = false
        if (_connectionState.value is ConnectionState.Scanning) {
            _connectionState.value = ConnectionState.Disconnected
        }
    }
    
    /**
     * Connect to a vehicle via BLE
     * 
     * @param address MAC address of the BLE device
     * @param autoReconnect Enable auto-reconnect on disconnect
     * @param callback Optional callback for connection status
     */
    fun connect(
        address: String,
        autoReconnect: Boolean = true,
        callback: ((Boolean) -> Unit)? = null
    ) {
        YuleDKCS.checkInitialized()
        
        // Stop scanning first
        stopScan()
        
        val bluetoothAdapter = android.bluetooth.BluetoothAdapter.getDefaultAdapter()
        val device = bluetoothAdapter?.getRemoteDevice(address)
        
        if (device == null) {
            callback?.invoke(false)
            return
        }
        
        connectionCallback = callback
        currentDeviceAddress = address
        _connectionState.value = ConnectionState.Connecting
        
        scope.launch {
            try {
                connect(device)
                    .useAutoConnect(autoReconnect)
                    .retry(3, 100)
                    .done {
                        Log.i(TAG, "Connected to $address")
                        _connectionState.value = ConnectionState.Connected(address)
                        connectionCallback?.invoke(true)
                        
                        // Request initial status
                        requestVehicleStatus()
                    }
                    .fail { _, status ->
                        Log.e(TAG, "Connection failed with status: $status")
                        _connectionState.value = ConnectionState.Error("连接失败: $status")
                        connectionCallback?.invoke(false)
                    }
                    .enqueue()
            } catch (e: Exception) {
                Log.e(TAG, "Error connecting", e)
                _connectionState.value = ConnectionState.Error(e.message ?: "Unknown error")
                callback?.invoke(false)
            }
        }
    }
    
    /**
     * Connect to nearest vehicle (highest RSSI)
     * 
     * @param callback Callback with Result containing device address
     */
    fun connectToNearest(callback: (Result<String>) -> Unit) {
        val nearest = _scanResults.value.firstOrNull()
        if (nearest == null) {
            callback(Result.failure(Exception("没有发现附近的车辆")))
            return
        }
        
        connect(nearest.device.address) { success ->
            if (success) {
                callback(Result.success(nearest.device.address))
            } else {
                callback(Result.failure(Exception("连接失败")))
            }
        }
    }
    
    /**
     * Disconnect from the current device
     */
    fun disconnect() {
        disconnect().enqueue()
        currentDeviceAddress = null
        _connectionState.value = ConnectionState.Disconnected
        _vehicleStatus.value = null
    }
    
    /**
     * Send a command to the connected vehicle
     * 
     * @param keyId The key identifier for authentication
     * @param command The command to send
     * @param callback Callback with result
     */
    fun sendCommand(
        keyId: String,
        command: Command,
        callback: (Result<CommandResponse>) -> Unit
    ) {
        YuleDKCS.checkInitialized()
        
        if (!isConnected) {
            callback(Result.failure(IllegalStateException("未连接到设备")))
            return
        }
        
        // Check permission
        if (!YuleDKCS.keyManager.hasPermission(keyId, command.toPermission())) {
            callback(Result.failure(SecurityException("钥匙没有此权限")))
            return
        }
        
        scope.launch {
            try {
                // Build command packet with encryption
                val commandData = buildCommandPacket(keyId, command)
                val commandId = System.currentTimeMillis().toString()
                
                // Store callback for response handling
                val responseDeferred = kotlinx.coroutines.CompletableDeferred<Result<ByteArray>>()
                pendingCommands[commandId] = { result ->
                    responseDeferred.complete(result)
                }
                
                // Send via BLE
                val data = Data(commandData)
                
                writeCharacteristic(
                    commandCharacteristic,
                    data,
                    BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
                ).await()
                
                // Wait for response with timeout
                val responseResult = withTimeoutOrNull(5000) {
                    responseDeferred.await()
                } ?: Result.failure(Exception("响应超时"))
                
                withContext(Dispatchers.Main) {
                    responseResult.fold(
                        onSuccess = { responseData ->
                            val cmdResponse = parseResponse(responseData)
                            
                            // Record usage log
                            YuleDKCS.keyManager.recordUsage(
                                keyId = keyId,
                                operation = command.toString(),
                                status = if (cmdResponse.success) 
                                    com.yuledkcs.model.KeyUsageLog.Status.SUCCESS 
                                else 
                                    com.yuledkcs.model.KeyUsageLog.Status.FAILURE,
                                failureReason = if (!cmdResponse.success) cmdResponse.message else null
                            )
                            
                            callback(Result.success(cmdResponse))
                        },
                        onFailure = { error ->
                            // Record failure
                            YuleDKCS.keyManager.recordUsage(
                                keyId = keyId,
                                operation = command.toString(),
                                status = com.yuledkcs.model.KeyUsageLog.Status.FAILURE,
                                failureReason = error.message
                            )
                            callback(Result.failure(error))
                        }
                    )
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error sending command", e)
                
                // Record failure
                YuleDKCS.keyManager.recordUsage(
                    keyId = keyId,
                    operation = command.toString(),
                    status = com.yuledkcs.model.KeyUsageLog.Status.FAILURE,
                    failureReason = e.message
                )
                
                withContext(Dispatchers.Main) {
                    callback(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Convenience method: Lock vehicle
     */
    fun lock(keyId: String, callback: (Result<CommandResponse>) -> Unit) {
        sendCommand(keyId, Command.Lock, callback)
    }
    
    /**
     * Convenience method: Unlock vehicle
     */
    fun unlock(keyId: String, callback: (Result<CommandResponse>) -> Unit) {
        sendCommand(keyId, Command.Unlock, callback)
    }
    
    /**
     * Convenience method: Start engine
     */
    fun startEngine(keyId: String, callback: (Result<CommandResponse>) -> Unit) {
        sendCommand(keyId, Command.StartEngine, callback)
    }
    
    /**
     * Convenience method: Stop engine
     */
    fun stopEngine(keyId: String, callback: (Result<CommandResponse>) -> Unit) {
        sendCommand(keyId, Command.StopEngine, callback)
    }
    
    /**
     * Convenience method: Open trunk
     */
    fun openTrunk(keyId: String, callback: (Result<CommandResponse>) -> Unit) {
        sendCommand(keyId, Command.TrunkOpen, callback)
    }
    
    /**
     * Convenience method: Open windows
     */
    fun openWindows(keyId: String, callback: (Result<CommandResponse>) -> Unit) {
        sendCommand(keyId, Command.OpenWindows, callback)
    }
    
    /**
     * Convenience method: Close windows
     */
    fun closeWindows(keyId: String, callback: (Result<CommandResponse>) -> Unit) {
        sendCommand(keyId, Command.CloseWindows, callback)
    }
    
    /**
     * Convenience method: Find vehicle (honk/flash)
     */
    fun findVehicle(keyId: String, callback: (Result<CommandResponse>) -> Unit) {
        sendCommand(keyId, Command.FindVehicle, callback)
    }
    
    /**
     * Request current vehicle status
     */
    fun requestVehicleStatus() {
        if (!isConnected) return
        
        scope.launch {
            try {
                val data = Data(byteArrayOf(CMD_GET_STATUS.toByte()))
                writeCharacteristic(commandCharacteristic, data, BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT).await()
            } catch (e: Exception) {
                Log.e(TAG, "Error requesting status", e)
            }
        }
    }
    
    /**
     * Check if currently connected to a device
     */
    fun isConnected(): Boolean = isConnected
    
    /**
     * Get current connected device address
     */
    fun getConnectedDevice(): String? = currentDeviceAddress
    
    override fun initialize() {
        // Set up notification for responses
        setNotificationCallback(responseCharacteristic)
            .with { _, data ->
                handleResponse(data.value)
            }
        
        enableNotifications(responseCharacteristic).enqueue()
        
        // Set up notification for status updates
        setNotificationCallback(statusCharacteristic)
            .with { _, data ->
                handleStatusUpdate(data.value)
            }
        
        enableNotifications(statusCharacteristic).enqueue()
    }
    
    override fun isRequiredServiceSupported(gatt: BluetoothGatt): Boolean {
        val service = gatt.getService(SERVICE_UUID)
        if (service != null) {
            commandCharacteristic = service.getCharacteristic(COMMAND_CHAR_UUID)
            responseCharacteristic = service.getCharacteristic(RESPONSE_CHAR_UUID)
            statusCharacteristic = service.getCharacteristic(STATUS_CHAR_UUID)
        }
        
        var writeRequest = false
        var notifyRequest = false
        
        commandCharacteristic?.let {
            val properties = it.properties
            writeRequest = properties and BluetoothGattCharacteristic.PROPERTY_WRITE > 0
        }
        
        responseCharacteristic?.let {
            val properties = it.properties
            notifyRequest = properties and BluetoothGattCharacteristic.PROPERTY_NOTIFY > 0
        }
        
        return commandCharacteristic != null && 
               responseCharacteristic != null &&
               writeRequest && notifyRequest
    }
    
    override fun onServicesInvalidated() {
        commandCharacteristic = null
        responseCharacteristic = null
        statusCharacteristic = null
    }
    
    private fun buildCommandPacket(keyId: String, command: Command): ByteArray {
        val key = YuleDKCS.keyManager.getKey(keyId)
            ?: throw IllegalStateException("钥匙未找到: $keyId")
        
        val timestamp = System.currentTimeMillis() / 1000
        val commandBytes = when (command) {
            is Command.Lock -> byteArrayOf(CMD_LOCK.toByte())
            is Command.Unlock -> byteArrayOf(CMD_UNLOCK.toByte())
            is Command.StartEngine -> byteArrayOf(CMD_START_ENGINE.toByte())
            is Command.StopEngine -> byteArrayOf(CMD_STOP_ENGINE.toByte())
            is Command.TrunkOpen -> byteArrayOf(CMD_OPEN_TRUNK.toByte())
            is Command.OpenWindows -> byteArrayOf(CMD_OPEN_WINDOWS.toByte())
            is Command.CloseWindows -> byteArrayOf(CMD_CLOSE_WINDOWS.toByte())
            is Command.FindVehicle -> byteArrayOf(CMD_FIND_VEHICLE.toByte())
            is Command.Custom -> command.data
        }
        
        // Build packet: [VERSION][KEY_ID][TIMESTAMP][COMMAND][SIGNATURE]
        val packet = ByteArray(1 + 16 + 8 + commandBytes.size + 64)
        var offset = 0
        
        // Version
        packet[offset++] = 0x01
        
        // Key ID (16 bytes)
        val keyIdBytes = keyId.toByteArray(Charsets.UTF_8)
        System.arraycopy(keyIdBytes, 0, packet, offset, minOf(keyIdBytes.size, 16))
        offset += 16
        
        // Timestamp (8 bytes, big-endian)
        for (i in 7 downTo 0) {
            packet[offset++] = (timestamp shr (i * 8)).toByte()
        }
        
        // Command data
        System.arraycopy(commandBytes, 0, packet, offset, commandBytes.size)
        offset += commandBytes.size
        
        // Sign packet using native crypto
        val dataToSign = packet.copyOfRange(0, offset)
        val signature = NativeLib.signCommand(dataToSign, key.keyData)
        
        System.arraycopy(signature, 0, packet, offset, signature.size)
        
        return packet
    }
    
    private fun handleResponse(data: ByteArray?) {
        data ?: return
        
        scope.launch(Dispatchers.Main) {
            if (data.size >= 2) {
                val status = data[0].toInt() and 0xFF
                val commandId = data[1].toInt() and 0xFF
                
                val callback = pendingCommands.remove(commandId.toString())
                
                if (status == 0x00) {
                    callback?.invoke(Result.success(data))
                } else {
                    callback?.invoke(Result.failure(Exception("命令执行失败: $status")))
                }
            }
        }
    }
    
    private fun handleStatusUpdate(data: ByteArray?) {
        data ?: return
        
        if (data.size >= 5) {
            val status = VehicleStatus(
                doorLocked = (data[0].toInt() and 0x01) != 0,
                engineRunning = (data[0].toInt() and 0x02) != 0,
                trunkOpen = (data[0].toInt() and 0x04) != 0,
                windowsOpen = (data[0].toInt() and 0x08) != 0,
                batteryLevel = data[1].toInt() and 0xFF,
                fuelLevel = data[2].toInt() and 0xFF,
                alarmActive = (data[0].toInt() and 0x10) != 0,
                timestamp = System.currentTimeMillis()
            )
            _vehicleStatus.value = status
        }
    }
    
    private fun parseResponse(data: ByteArray): CommandResponse {
        return if (data.isNotEmpty() && data[0].toInt() == 0x00) {
            CommandResponse(
                success = true,
                message = "成功"
            )
        } else {
            CommandResponse(
                success = false,
                message = "失败: ${data.getOrNull(0)?.toInt()}"
            )
        }
    }
    
    private fun Command.toPermission(): com.yuledkcs.model.Permission {
        return when (this) {
            is Command.Lock -> com.yuledkcs.model.Permission.LOCK
            is Command.Unlock -> com.yuledkcs.model.Permission.UNLOCK
            is Command.StartEngine -> com.yuledkcs.model.Permission.START_ENGINE
            is Command.StopEngine -> com.yuledkcs.model.Permission.STOP_ENGINE
            is Command.TrunkOpen -> com.yuledkcs.model.Permission.OPEN_TRUNK
            is Command.OpenWindows -> com.yuledkcs.model.Permission.OPEN_WINDOWS
            is Command.CloseWindows -> com.yuledkcs.model.Permission.CLOSE_WINDOWS
            is Command.FindVehicle -> com.yuledkcs.model.Permission.FIND_VEHICLE
            is Command.Custom -> com.yuledkcs.model.Permission.UNLOCK
        }
    }
    
    data class CommandResponse(
        val success: Boolean,
        val message: String
    )
}

// Extension for timeout
private suspend inline fun <T> withTimeoutOrNull(timeMillis: Long, block: () -> T): T? {
    return try {
        kotlinx.coroutines.withTimeout(timeMillis) {
            block()
        }
    } catch (e: kotlinx.coroutines.TimeoutCancellationException) {
        null
    }
}
